internal/engine/procedures/typing/typing.go

```diff
@@ -0,0 +1,663 @@
+	if ok {
+		argsTypes := make([]*types.DataType, len(p0.Arguments))
+		for i, arg := range p0.Arguments {
+			argsTypes[i] = arg.Accept(t).(*types.DataType)
+		}
+
+		returnType, err := funcDef.ValidateArgs(argsTypes)
+		if err != nil {
+			panic(err)
+		}
+
+		return &types.ProcedureReturn{
+			Fields: []*types.NamedType{
+				{
+					Type: returnType,
+				},
+			},
+		}
+	}
+
+	// check if it is a procedure
+	var proc *types.Procedure
+	found := false
+	for _, p := range t.currentSchema.Procedures {
+		if p.Name == p0.Name {
+			proc = p
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		panic(fmt.Sprintf(`unknown function/procedure "%s"`, p0.Name))
+	}
+
+	// check the args
+	if len(p0.Arguments) != len(proc.Parameters) {
+		panic(fmt.Sprintf(`expected %d arguments, got %d`, len(proc.Parameters), len(p0.Arguments)))
+	}
+
+	for i, arg := range p0.Arguments {
+		argType := arg.Accept(t).(*types.DataType)
+		if !argType.Equals(proc.Parameters[i].Type) {
+			panic(fmt.Sprintf(`argument %d has wrong type`, i))
+		}
+	}
+
+	if proc.Returns == nil {
+		return &types.ProcedureReturn{} // avoid nil pointer
+	}
+
+	return proc.Returns
+}
+
+func (t *typingVisitor) VisitStatementBreak(p0 *parser.StatementBreak) any {
+	return nil
+}
+
+func (t *typingVisitor) VisitStatementForLoop(p0 *parser.StatementForLoop) any {
+	t.loopTarget = p0.Variable
+
+	switch target := p0.Target.(type) {
+	case *parser.LoopTargetVariable:
+		r := target.Accept(t).(*types.DataType)
+
+		// we will not declare these as anonymous, since this is simply
+		// a field in an array of a known type
+		_, ok := t.declarations[p0.Variable]
+		if ok {
+			panic("variable already declared")
+		}
+
+		t.declarations[p0.Variable] = r
+	case *parser.LoopTargetCall:
+		// this can be either a table or a custom data type
+		r := target.Accept(t).(map[string]*types.DataType)
+		_, ok := t.anonymousDeclarations[p0.Variable]
+		if ok {
+			panic("variable already declared")
+		}
+
+		t.anonymousDeclarations[p0.Variable] = r
+	case *parser.LoopTargetRange:
+		r := target.Accept(t).(*types.DataType) // int
+
+		// we will not declare these as anonymous, since it is a simple int
+		_, ok := t.declarations[p0.Variable]
+		if ok {
+			panic("variable already declared")
+		}
+
+		t.declarations[p0.Variable] = r
+	case *parser.LoopTargetSQL:
+		r := target.Accept(t).(map[string]*types.DataType)
+
+		_, ok := t.anonymousDeclarations[p0.Variable]
+		if ok {
+			panic("variable already declared")
+		}
+
+		t.anonymousDeclarations[p0.Variable] = r
+	default:
+		panic("unknown loop target")
+	}
+
+	for _, stmt := range p0.Body {
+		stmt.Accept(t)
+	}
+
+	t.loopTarget = ""
+
+	return nil
+}
+
+// analyzeSQL analyzes the given SQL statement and returns the resulting relation.
+func (t *typingVisitor) analyzeSQL(stmt tree.AstNode) *engine.Relation {
+	// TODO: we have a problem here where the sql analyzer cannot analyze
+	// the @ vars, since it is using the current_setting() command.
```

Comment:
Is this a problem because we run the tree through the cleaner before typing?  I guess we can't reverse that order?
And the impact of the problem is that we just can't check that type, but it otherwise doesn't break anything?

Reply:
Correct. Still working on a workaround for this.